compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -538,7 +538,8 @@ class Namer { typer: Typer =>
             trees.map {
               case mcls @ TypeDef(name, impl: Template) if valid(mcls) =>
                 moduleClsDef.get(name) match {
-                  case Some((stat1, mcls1@TypeDef(_, impl1: Template))) =>
+                  case Some((stat1, mcls1)) =>
+                    assert(mcls1.rhs.isInstanceOf[Template])
```

Comment:
I think I prefer the old version. It's shorter and cleaner. You get the value that did not match so that should be enough to track it down.